Implement `orphans` and `widows` computed values

rwlbuis opened the pull request:
#### f1e11b2d0edbc78fdcd8f41e10bd345a2d95940f
<pre>
Implement `orphans` and `widows` computed values
<a href="https://bugs.webkit.org/show_bug.cgi?id=299374">https://bugs.webkit.org/show_bug.cgi?id=299374</a>

Reviewed by Sam Weinig.

The `orphans` and `widows` computed values should not reflect the internal auto value.

* LayoutTests/editing/pasteboard/data-transfer-get-data-on-drop-rich-text-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-without-renderer-expected.txt:
* LayoutTests/fast/events/before-input-events-prevent-drag-and-drop-expected.txt:
* LayoutTests/fast/events/input-events-paste-rich-datatransfer-expected.txt:
* LayoutTests/fast/events/ondrop-text-html-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-break/inheritance-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
* LayoutTests/platform/glib/fast/events/input-events-paste-rich-datatransfer-expected.txt:
* LayoutTests/platform/mac/fast/events/ondrop-text-html-expected.txt:
* LayoutTests/svg/css/getComputedStyle-basic-expected.txt:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/style/StyleExtractorCustom.h:
(WebCore::Style::PropertyExtractorAdaptor&lt;CSSPropertyWidows&gt;::computedValue const):
(WebCore::Style::PropertyExtractorAdaptor&lt;CSSPropertyOrphans&gt;::computedValue const):
(WebCore::Style::ExtractorCustom::extractWidows):
(WebCore::Style::ExtractorCustom::extractOrphans):
(WebCore::Style::ExtractorCustom::extractWidowsSerialization):
(WebCore::Style::ExtractorCustom::extractOrphansSerialization):

Canonical link: <a href="https://commits.webkit.org/300690@main">https://commits.webkit.org/300690@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8fe4619f4c3254149d99d490eebb743345dc11bc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123517 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43232 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33928 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130239 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75658 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125394 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43955 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51826 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93901 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62330 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126470 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35019 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110503 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74521 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33990 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28660 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73754 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104736 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28885 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132960 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50468 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38419 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102397 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50843 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106724 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102242 "Found 2 new API test failures: WebKitGTK/TestWebsiteData:/webkit/WebKitWebsiteData/itp, WebKitGTK/TestLoaderClient:/webkit/WebKitWebPage/get-uri (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25986 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47591 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25821 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/47276 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50322 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56083 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49796 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53143 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51471 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/76a64cc9ba3bdd5a3ca78e23328afe0b0834fa2e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51197

rwlbuis commented on Source/WebCore/style/StyleExtractorCustom.h:
```diff
@@ -3329,5 +3343,15 @@ inline void ExtractorCustom::extractWebkitMaskPositionShorthandSerialization(Ext
+inline void ExtractorCustom::extractOrphansSerialization(ExtractorState& state, StringBuilder& builder, const CSS::SerializationContext& context)
+{
+    ExtractorSerializer::serializeNumber(state, builder, context, state.style.orphans().tryValue().value_or(2).value);
+}
```
Sure, done.

rwlbuis commented:
> Just so the question doesn't get lost:
> 
> @rwlbuis:
> 
> > What is the reason we have the auto state at all? Do we get rendering wrong in that case (as presumably it behaves different that if you passed in 2)?

I haven't looked at the details in a while, but I think it is done so internally we can use some different initial values (0, 1 or both). I tried removing the 'auto' at some point and it was hard to do and I gave up. I think @ntim tried the same, there should be an existing bug for it, @ntim may know more.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/61c517f81947882d624000eaeab2572ccd8305a1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51197

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8fe4619f4c3254149d99d490eebb743345dc11bc. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51197

rwlbuis committed:
Implement `orphans` and `widows` computed values
https://bugs.webkit.org/show_bug.cgi?id=299374

Reviewed by Sam Weinig.

The `orphans` and `widows` computed values should not reflect the internal auto value.

* LayoutTests/editing/pasteboard/data-transfer-get-data-on-drop-rich-text-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-without-renderer-expected.txt:
* LayoutTests/fast/events/before-input-events-prevent-drag-and-drop-expected.txt:
* LayoutTests/fast/events/input-events-paste-rich-datatransfer-expected.txt:
* LayoutTests/fast/events/ondrop-text-html-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-break/inheritance-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
* LayoutTests/platform/glib/fast/events/input-events-paste-rich-datatransfer-expected.txt:
* LayoutTests/platform/mac/fast/events/ondrop-text-html-expected.txt:
* LayoutTests/svg/css/getComputedStyle-basic-expected.txt:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/style/StyleExtractorCustom.h:
(WebCore::Style::PropertyExtractorAdaptor<CSSPropertyWidows>::computedValue const):
(WebCore::Style::PropertyExtractorAdaptor<CSSPropertyOrphans>::computedValue const):
(WebCore::Style::ExtractorCustom::extractWidows):
(WebCore::Style::ExtractorCustom::extractOrphans):
(WebCore::Style::ExtractorCustom::extractWidowsSerialization):
(WebCore::Style::ExtractorCustom::extractOrphansSerialization):

Canonical link: https://commits.webkit.org/300690@main
